libraries/standard/mqtt/src/mqtt.c

```diff
@@ -951,10 +965,30 @@ static MQTTStatus_t receiveConnack( const MQTTContext_t * pContext,
                                                       pContext->transportInterface.networkContext,
                                                       pIncomingPacket );
 
-        timeExpired = ( calculateElapsedTime( getTimeStamp(), entryTimeMs ) >= timeoutMs ) ? true : false;
+        /* The loop times out based on 2 conditions.
+         * 1. If timeoutMs is greater than 0:
+         *    Loop times out based on the timeout calculated by getTime()
+         *    function.
+         * 2. If timeoutMs is 0:
+         *    Loop times out based on the maximum number of retries config
+         *    MQTT_MAX_CONNACK_RECEIVE_RETRY_COUNT. This config will control
+         *    maximum the number of retry attempts to read the CONNACK packet.
+         *    A value of 0 for the config will try once to read CONNACK. */
+        useRetry = ( timeoutMs == 0U ) ? true : false;
+
+        if( useRetry == false )
+        {
+            /* Loop runs only once for 0 timeoutMs. */
+            breakFromLoop = ( calculateElapsedTime( getTimeStamp(), entryTimeMs ) >= timeoutMs ) ? true : false;
+        }
+        else
+        {
+            breakFromLoop = ( loopCount >= MQTT_MAX_CONNACK_RECEIVE_RETRY_COUNT ) ? true : false;
+            loopCount++;
+        }
 
-        /* Loop until there is data to read or if the timeout has not expired. */
-    } while( ( status == MQTTNoDataAvailable ) && ( timeExpired != true ) );
+        /* Loop until there is data to read or if we have exceeded the timeout/retries. */
```

Comment:
Should there be some delay or back off logic in between consecutive transport receive calls?

Reply:
There is already a delay as part of the transport interface configured by the application. So the transport read is not going to immediately return, but wait till the timeout is hit if there is no data to read. I think adding a back off and retry may not be required here as library retrying may be enough as CONNACK is going to be the only packet being received.